kubeai: document usage of NRI plugins for performance optimization

marquiz opened the pull request:
Document a preferred setup of the Balloons Policy from the NRI Plugins project.

eero-t commented on kubeai/README.md:
```diff
@@ -181,6 +181,109 @@ curl "http://localhost:8000/openai/v1/chat/completions" \
+
+## Installation of Balloons Policy Plugin
+
+> **NOTE:** In order to prevent disturbing already running workloads it is
+> recommended to install the NRI plugin to an empty node (right after node
+> bootstrap or drain the node before installation).
```
```suggestion
> **NOTE:** To avoid disturbing already running workloads it is
> recommended to install the NRI plugin to an empty node (do it right after node
> bootstrap, or to drain the node before installation).
```

marquiz replied:
Thx, fixed

eero-t commented on kubeai/README.md:
````diff
@@ -181,6 +181,109 @@ curl "http://localhost:8000/openai/v1/chat/completions" \
+helm install -n kube-system balloons nri-plugins/nri-resource-policy-balloons
+```
+
+> **NOTE**: If using containerd version earlier than v2.0 you need to enable
+> the NRI support in the containerd configuration file. Instead of manual
+> configuration you can provide `--set nri.runtime.patchConfig=true` to the Helm
+> command above, which will automatically patch the containerd configuration
+> file on each node.
````
```suggestion
> **NOTE**: With `containerd` version earlier than v2.0 you need to enable
> the NRI support in the `containerd` configuration file. Instead of manual
> configuration you can provide `--set nri.runtime.patchConfig=true` to the Helm
> command above, which will automatically patch the `containerd` configuration
> file on each node.
```

marquiz replied:
Thx. Did s/If using/With/.

However I didn't put containerd in backticks as referring to the containerd project/package rather than a shell command directly. Dunno if there is some OPEA style guide, probably not :smile:  Thoughts?

poussa replied:
this is fine.

eero-t replied:
Ah, it's officially written with lower-case `c`.  One option would be to use italics for names, but this is fine as-is.

eero-t commented on kubeai/README.md:
````diff
@@ -181,6 +181,109 @@ curl "http://localhost:8000/openai/v1/chat/completions" \
+```
+
+The configuration above allocates at minimum 12 logical CPUs (6 full CPU cores
+assuming that hyperthreading is enabled) for the inference serving pods. With
````
Pods, not containers?
```suggestion
assuming that hyperthreading is enabled) for the vLLM pods. With
```

marquiz replied:
I heavily rephrased this paragraph. No reference to vLLM anymore (just "inference")

eero-t commented on kubeai/README.md:
```diff
@@ -181,6 +181,109 @@ curl "http://localhost:8000/openai/v1/chat/completions" \
+The configuration above allocates at minimum 12 logical CPUs (6 full CPU cores
+assuming that hyperthreading is enabled) for the inference serving pods. With
+the settings of the `kubeai-vllm` balloon it creates a new dedicated balloon
+for each inference engine container as long as there are enough free cpus on
```
```suggestion
for each inference engine container as long as there are enough free CPUs on
```

marquiz replied:
Fixed

eero-t reviewed (commented):
README is getting quite long now, so I wonder would this make sense as a separate document (linked from README), as it's more about configuring the cluster, than configuring KubeAI?

marquiz commented on kubeai/README.md:
```diff
@@ -181,6 +181,109 @@ curl "http://localhost:8000/openai/v1/chat/completions" \
+## Configuration of Balloons Policy Plugin
+
+The aim of the balloons policy configuration is to isolate the vllm containers
+to minimize their interference with other containers.
```
Thx, changed. Except for vllm as now I'm referring to inference in more generic terms (added support for ollama, too)

marquiz commented:
Updated:
- marked ready-for-review
- added support for ollama
- removed `minBalloons` and `minCPUs` settings
- added (optional) `showContainersInNrt` plus `agent.nodeResourceTopology` for debugging/investigation

poussa commented:
> README is getting quite long now, so I wonder would this make sense as a separate document (linked from README), as it's more about configuring the cluster, than configuring KubeAI?

Let's keep it in single file. There is room to trim the README to make it more compact.

eero-t commented on kubeai/README.md:
```diff
@@ -181,6 +185,120 @@ curl "http://localhost:8000/openai/v1/chat/completions" \
+containers by hiding hyperthreads from them. For example, if a pod requests 6
+CPUs, the balloon will reserve 6 full physical CPU cores (i.e., 12 logical
+CPUs), but configure the cpuset so that the inference instance can only use 6
+logical CPUs—one per core—while the sibling hyperthreads remain unused. With
+this configuration for the `kubeai-inference` balloon, a new dedicated balloon
+is created for each inference engine container, as long as enough free CPUs are
```
Very long paragraph, split for readability:
```suggestion
logical CPUs—one per core—while the sibling hyperthreads remain unused.

Above `kubeai-inference` balloon configuration craetes a new dedicated balloon
for each inference engine container, as long as enough free CPUs are
```

marquiz replied:
The paragraph is not super long. The sentences maybe. Reworded but still in one paragraph.

marquiz commented on kubeai/README.md:
```diff
@@ -181,6 +185,120 @@ curl "http://localhost:8000/openai/v1/chat/completions" \
+## Configuration of Balloons Policy Plugin
+
+The aim of the balloons policy configuration is to isolate the model (inference
+serving) containers to minimize the impact of containers on each other.
```
Changed as sugggested

marquiz commented on kubeai/README.md:
```diff
@@ -181,6 +185,120 @@ curl "http://localhost:8000/openai/v1/chat/completions" \
+    prometheusExport: true
+  # NodeResourceTopology custom resource objects can be enabled for debugging
+  # and closely # examining the CPU pinning of inference containers (not
+  # recommended for # production # because of potentially excessive load on the
```
Changed/fixed

marquiz commented on kubeai/README.md:
````diff
@@ -181,6 +185,120 @@ curl "http://localhost:8000/openai/v1/chat/completions" \
+EOF
+```
+
+The configuration above allocates full CPU cores to inference serving
````
Check, changed as suggested

askervin commented:
@marquiz, I needed to change

```
key: labels/app.kubernetes.io/name
```
to
```
key: pod/labels/app.kubernetes.io/name
```

when otherwise label match was not working, and the policy resolved vllm container to be put in the default balloon type. Logs looked like this:
```
I0611 14:16:53.510691       1 log.go:476] D: [           policy           ] choosing balloon type for container default/docsum-vllm-7b787d6cf7-fqwwz/vllm...
I0611 14:16:53.510722       1 log.go:476] D: [           policy           ] - checking expression <labels/app.kubernetes.io/name In vllm,ollama> of balloon type "kubeai-inference" against container default/docsum-vllm-7b787d6cf7-fqwwz/vllm...
I0611 14:16:53.510742       1 log.go:476] D: [           policy           ] - checking expression <name In server> of balloon type "kubeai-inference" against container default/docsum-vllm-7b787d6cf7-fqwwz/vllm...
I0611 14:16:53.510775       1 log.go:476] D: [           policy           ] - namespace "default" matches namespaces of balloon type "default"
```

When prefixed with "pods/", the expression resulted in:

```
checking expression <pod/labels/app.kubernetes.io/name In vllm,ollama> of balloon type "kubeai-inference" against container default/docsum-vllm-7b787d6cf7-4nths/vllm...
=> matches
```

in the log, and NRT looked as expected:

kubectl get noderesourcetopologies.topology.node.k8s.io -o yaml
```
...
  - attributes:
    - name: cpuset
      value: 86-93,258-265
    - name: shared cpuset
      value: ""
    - name: excess cpus
      value: 0m
    name: kubeai-inference[0]
    resources:
    - allocatable: "342"
      available: "326"
      capacity: "344"
      name: cpu
    type: balloon
  - attributes:
    - name: cpuset
      value: 86-93
    - name: memory set
      value: ""
    name: default/docsum-vllm-7b787d6cf7-4nths/vllm
    parent: kubeai-inference[0]
    resources:
    - allocatable: "16"
      available: "0"
      capacity: "8"
      name: cpu
    type: allocation for container
...
```

I think this is a usability issue in our expression evaluation. It would be reasonable to expect "labels/" automatically match pod labels. What do you think?

marquiz commented:
> It would be reasonable to expect "labels/" automatically match pod labels. What do you think?

Not so sure about this as there are also container labels (in CRI/NRI) level. I'd leave that as is and possibly only note that caveat in the documentation.
